src/trajectory/util.cpp

```diff
@@ -0,0 +1,320 @@
+aikido::trajectory::UniqueSplinePtr createPartialTrajectory(
+    const aikido::trajectory::Spline& traj, double partialStartTime)
+{
+  if (partialStartTime < traj.getStartTime()
+      || partialStartTime > traj.getEndTime())
+    throw std::runtime_error("Wrong partial start time");
+
+  auto stateSpace = traj.getStateSpace();
+  const std::size_t dimension = stateSpace->getDimension();
+  auto outputTrajectory = make_unique<aikido::trajectory::Spline>(
+      stateSpace, traj.getStartTime());
+
+  double currSegmentStartTime = traj.getStartTime();
+  double currSegmentEndTime = currSegmentStartTime;
+  std::size_t currSegmentIdx = 0;
+
+  auto segmentStartState = stateSpace->createState();
+  auto segmentEndState = stateSpace->createState();
+  Eigen::VectorXd segStartPos(dimension);
+  Eigen::VectorXd segEndPos(dimension);
+  Eigen::VectorXd segStartVel(dimension);
+  Eigen::VectorXd segEndVel(dimension);
+  const Eigen::VectorXd zeroPos = Eigen::VectorXd::Zero(dimension);
+  traj.evaluate(partialStartTime, segmentStartState);
+  stateSpace->logMap(segmentStartState, segStartPos);
+  traj.evaluateDerivative(partialStartTime, 1, segStartVel);
+
+  while (currSegmentIdx < traj.getNumSegments())
+  {
+    currSegmentEndTime += traj.getSegmentDuration(currSegmentIdx);
+    if (partialStartTime >= currSegmentStartTime
+        && partialStartTime <= currSegmentEndTime)
+    {
+      // create new segment
+      traj.evaluate(currSegmentEndTime, segmentEndState);
+      stateSpace->logMap(segmentEndState, segEndPos);
+      traj.evaluateDerivative(currSegmentEndTime, 1, segEndVel);
+
+      double segmentDuration = currSegmentEndTime - partialStartTime;
+
+      if (segmentDuration > 0.0)
+      {
+        CubicSplineProblem problem(
+            Eigen::Vector2d{0., segmentDuration}, 4, dimension);
+        problem.addConstantConstraint(0, 0, zeroPos);
+        problem.addConstantConstraint(0, 1, segStartVel);
+        problem.addConstantConstraint(1, 0, segEndPos - segStartPos);
+        problem.addConstantConstraint(1, 1, segEndVel);
+        const auto solution = problem.fit();
+        const auto coefficients = solution.getCoefficients().front();
+        outputTrajectory->addSegment(
+            coefficients, segmentDuration, segmentStartState);
+      }
+      break;
+    }
+
+    currSegmentIdx++;
+    currSegmentStartTime = currSegmentEndTime;
+  }
+
+  for (std::size_t i = currSegmentIdx + 1; i < traj.getNumSegments(); i++)
+  {
+    outputTrajectory->addSegment(
+        traj.getSegmentCoefficients(i),
+        traj.getSegmentDuration(i),
+        traj.getSegmentState(i));
+  }
+  return outputTrajectory;
+}
+
+//==============================================================================
+aikido::trajectory::UniqueInterpolatedPtr convertToInterpolated(
+    const aikido::trajectory::Spline& traj,
+    aikido::statespace::ConstInterpolatorPtr& interpolator)
```

Comment:
Pass `interpolator` by value and move it to the constructor of `Interpolated` as:
```cpp
  auto outputTrajectory
      = make_unique<aikido::trajectory::Interpolated>(stateSpace, std::move(interpolator));
```

The rationale is as follows:
1. No pass by reference: We don't change the pointer (but share the ownership).
2. Pass by reference: This is helpful to indicate that this function takes the ownership.

FYI, here is the [convention](https://github.com/personalrobotics/aikido/blob/master/STYLE.md#smart-pointers).

I understand this is a bit subtle, but let's try to be consistent with other code.